test/extended/util/adminack.go

```diff
@@ -0,0 +1,243 @@
+package util
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configv1client "github.com/openshift/client-go/config/clientset/versioned"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+// AdminAckTest contains artifacts used during test
+type AdminAckTest struct {
+	Oc     *CLI
+	Config *restclient.Config
+}
+
+const adminAckGateFmt string = "^ack-[4-5][.]([0-9]{1,})-[^-]"
+
+var adminAckGateRegexp = regexp.MustCompile(adminAckGateFmt)
+
+// Test simply returns successfully if admin ack functionality is not part the baseline being tested. Otherwise,
+// for each configured admin ack gate, test verifies the gate name format and that it contains a description. If
+// valid and the gate is applicable to the OCP version under test, test checks the value of the admin ack gate.
+// If the gate has been ack'ed the test verifies that Upgradeable condition is true and clears the ack. Test then
+// verifies Upgradeable condition is false and contains correct reason and correct message. It then modifies the
+// admin-acks configmap to ack the given admin-ack gate. Once all gates have been ack'ed, the test waits for the
+// Upgradeable condition to change to true.
+func (t *AdminAckTest) Test(ctx context.Context) {
+
+	// Check if this release has admin ack functionality.
+	gateCm, errMsg := getAdminGatesConfigMap(ctx, t.Oc)
+	if len(errMsg) != 0 {
+		framework.Failf(errMsg)
+	}
+	if gateCm == nil || (gateCm != nil && len(gateCm.Data) == 0) {
+		framework.Logf("Skipping admin ack test. Admin ack is not in this baseline or contains no gates.")
+		return
+	}
+	ackCm, errMsg := getAdminAcksConfigMap(ctx, t.Oc)
+	if len(errMsg) != 0 {
+		framework.Failf(errMsg)
+	}
+	currentVersion := getCurrentVersion(ctx, t.Config)
+	var msg string
+	for k, v := range gateCm.Data {
+		ackVersion := adminAckGateRegexp.FindString(k)
+		if ackVersion == "" {
+			framework.Failf(fmt.Sprintf("Configmap openshift-config-managed/admin-gates gate %s has invalid format; must comply with %q.", k, adminAckGateFmt))
+		}
+		if v == "" {
+			framework.Failf(fmt.Sprintf("Configmap openshift-config-managed/admin-gates gate %s does not contain description.", k))
+		}
+		if !gateApplicableToCurrentVersion(ackVersion, currentVersion) {
+			continue
+		}
+		if ackCm.Data[k] == "true" {
+			if upgradeableExplicitlyFalse(ctx, t.Config) {
+				if adminAckRequiredWithMessage(ctx, t.Config, v) {
+					framework.Failf(fmt.Sprintf("Gate %s has been ack'ed but Upgradeable is "+
+						"false with reason AdminAckRequired and message %q.", k, v))
+				}
+				framework.Logf(fmt.Sprintf("Gate %s has been ack'ed. Upgradeable is "+
+					"false but not for reason AdminAckRequired.", k) + " " + getUpgradeable(ctx, t.Config))
```

Comment:
Should be `but not for reason AdminAckRequired mentioning this key.`, because we might be `AdminAckRequired` on other keys.